tests/net/tcp/src/main.c

```diff
@@ -2856,4 +2861,145 @@ ZTEST(net_tcp, test_client_seq_validation)
 	k_sleep(K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
 }
 
+struct ack_check_struct {
+	int s_data_len;
+	int c_ack;
+	int c_exp_seq;
+};
+
+/* The Congestion Window is by default enabled to be small (e.g. 67B),
+ * so DONT set the s_data_len bigger than it or the expected_seq will
+ * fail the check.
+ */
+static struct ack_check_struct ack_check_list[] = {
+	{10, 20, 10},
+	{20, 9, 30},
+	{30, -2000, 60},
+	{50, 111, 110},
+};
+
+static int ack_check_index;
+static uint32_t expected_seq;
+static uint32_t svr_seq_base;
+static int ackerr;
+
+static void handle_server_ack_validation_test(struct net_pkt *pkt)
+{
+	struct tcphdr th;
+	struct net_pkt *reply;
+	int ret;
+
+	ret = read_tcp_header(pkt, &th);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	if (FL(&th.th_flags, &, PSH)) {
+		/* If server data packet, then return with c_ack in the ack_check_list[] */
+		uint32_t old_ack = ack;
+
+		ack = (uint32_t)(svr_seq_base + ack_check_list[ack_check_index].c_ack);
+		reply = prepare_ack_packet(net_pkt_family(pkt), htons(MY_PORT), htons(PEER_PORT));
+		zassert_not_null(reply, "Cannot create pkt");
+		ack = old_ack;
+
+		expected_seq = (uint32_t)(svr_seq_base + ack_check_list[ack_check_index].c_exp_seq);
+		ackerr = GET_STAT(net_iface, tcp.ackerr);
+
+		ret = net_recv_data(net_iface, reply);
+		zassert_true(ret == 0, "recv data failed (%d)", ret);
+
+		/* Let the IP stack to process the packet properly */
+		k_yield();
+	} else {
+		int ackerr_after;
+
+		/* For all non-data packets, we assume it is ACK-only because we use RST to
+		 * close the TCP connection later.
+		 */
+		test_verify_flags(&th, ACK);
+		/* Verify that we received the ACK-only packet with the expected seqnum */
+		zassert_equal(expected_seq, ntohl(th.th_seq),
+			      "Wrong seqnum received from server. "
+			      "Expected SEQ %u but got %u",
+			      expected_seq, ntohl(th.th_seq));
+
+		ackerr_after = GET_STAT(net_iface, tcp.ackerr);
+		zassert_equal(ackerr + 1, ackerr_after,
+			      "Wrong ACK number not detected (before %d, after %d)", ackerr,
+			      ackerr_after);
+
+		/* send valid ACK to server to avoid retransmission */
+		ack += (uint32_t)ack_check_list[ack_check_index].s_data_len;
```

Comment:
Same here, the s_data_len could be uint32_t especially as the value probably cannot be negative anyway